File: revision/8_may_24/tree_ds/heap.cpp
#include <bits/stdc++.h>
#include <cassert>
#include <cstdio>
#include <vector>
using namespace std;

class MaxHeap{
    private:
        vector<int> bucket;
        void move_up(int i){
            int value = bucket[i];
            int child = i;
            int parent = (child-1)/2;
            while (child > 0 && value > bucket[parent]){
                bucket[child] = bucket[parent];
                child = parent;
                parent = (child-1)/2;
            }
            bucket[child] = value;
        }

        void move_down(int i){
            int n = bucket.size();
            int value = bucket[i];
            int parent = i;
            int left_child = 2*parent + 1;
            int right_child = 2*parent + 2;
            while (left_child < n){
                if (right_child < n){
                    if (value < bucket[left_child] && value < bucket[right_child]){
                        if (bucket[left_child] > bucket[right_child]){
                            bucket[parent] = bucket[left_child];
                            parent = left_child;
                        }else {
                            bucket[parent] = bucket[right_child];
                            parent = right_child; 
                        }
                    }else if (value < bucket[left_child]) {
                        bucket[parent] = bucket[left_child];
                        parent = left_child;
                    }else if (value < bucket[right_child]) {
                        bucket[parent] = bucket[right_child];
                        parent = right_child;
                    }break;
                }else if (bucket[left_child] > value) {
                    bucket[parent] = bucket[left_child];
                    parent = left_child;
                }else break;
                left_child = 2*parent + 1;
                right_child = 2*parent + 2;
            }
            bucket[parent] = value;
        }

    public:
        MaxHeap(): bucket() {}
        MaxHeap(const vector<int>& vec){
            for (const int& v: vec)bucket.push_back(v);
            for (int i=1;i<bucket.size();i++)move_up(i);
        }

        void push(int value){            
            bucket.push_back(value);
            move_up(bucket.size()-1);
        }

        void pop(){
            int value = bucket.back();
            bucket.pop_back();
            bucket[0] = value;
            move_down(0);
        }

        int top() const {
            return bucket[0];
        }

        bool empty() const {
            return bucket.empty();
        }
};

int main(){
    MaxHeap heap{};
    assert(heap.empty());
    heap.push(1);
    assert(!heap.empty());
    assert(heap.top() == 1);

    heap.push(5);assert(heap.top() == 5);
    heap.push(2);assert(heap.top() == 5);
    heap.push(9);assert(heap.top() == 9);
    heap.push(12);assert(heap.top() == 12);
    heap.push(7);assert(heap.top() == 12);

    heap.pop();assert(heap.top() == 9);
    heap.pop();assert(heap.top() == 7);
    heap.pop();assert(heap.top() == 5);
    heap.pop();assert(heap.top() == 2);
    heap.pop();assert(heap.top() == 1);

    MaxHeap heap2{vector<int> {1,5,2,9,12,7}};
    assert(heap2.top() == 12);
    heap2.pop();assert(heap2.top() == 9);
    heap2.pop();assert(heap2.top() == 7);
    heap2.pop();assert(heap2.top() == 5);
    heap2.pop();assert(heap2.top() == 2);
    heap2.pop();assert(heap2.top() == 1);

}

File: revision/29_jun_24/src/tree/heap.rs
struct Heap {
    inner: Vec<i32>,
}

impl Heap {
    fn new() -> Self {
        Self { inner: Vec::new() }
    }

    fn move_up(v: &mut [i32], index: usize) {
        let value = v[index];
        let mut child = index as i32;
        let mut parent = (child - 1) / 2;
        while child > 0 && value > v[parent as usize] {
            v[child as usize] = v[parent as usize];
            child = parent;
            parent = (child - 1) / 2
        }
        v[child as usize] = value;
    }

    fn move_down(v: &mut [i32], size: usize) {
        let size = size as i32;
        let value = v[0];
        let mut parent: i32 = 0;
        let mut left_child = 2 * parent + 1;
        let mut right_child = 2 * parent + 2;
        loop {
            if left_child <= size - 1 && right_child <= size - 1 {
                if v[left_child as usize] > value && v[right_child as usize] > value {
                    if v[left_child as usize] > v[right_child as usize] {
                        v[parent as usize] = v[left_child as usize];
                        parent = left_child;
                    } else {
                        v[parent as usize] = v[right_child as usize];
                        parent = right_child;
                    }
                } else if v[left_child as usize] > value {
                    v[parent as usize] = v[left_child as usize];
                    parent = left_child;
                } else if v[right_child as usize] > value {
                    v[parent as usize] = v[right_child as usize];
                    parent = right_child;
                } else {
                    break;
                }
            } else if left_child <= size - 1 && v[left_child as usize] > value {
                v[parent as usize] = v[left_child as usize];
                parent = left_child;
            } else if right_child <= size - 1 && v[right_child as usize] > value {
                v[parent as usize] = v[right_child as usize];
                parent = right_child;
            } else {
                break;
            }
            left_child = 2 * parent + 1;
            right_child = 2 * parent + 2;
        }
        v[parent as usize] = value;
    }

    fn is_empty(&self) -> bool {
        self.inner.len() == 0
    }

    fn push(&mut self, elem: i32) {
        self.inner.push(elem);
        let index = self.inner.len() - 1;
        Heap::move_up(&mut self.inner, index);
    }

    fn pop(&mut self) {
        if self.is_empty() {
            panic!("heap is empty")
        }
        self.inner[0] = *self.inner.last().unwrap();
        self.inner.pop();
        let size = self.inner.len();
        Heap::move_down(&mut self.inner, size);
    }

    fn top(&self) -> i32 {
        if self.is_empty() {
            panic!("heap is empty");
        }
        self.inner[0]
    }

    // use heap_sort to sort
    // provided vector `a`
    fn sort(v: &mut Vec<i32>) {
        for i in 1..v.len() {
            Heap::move_up(v, i);
        }
        let mut size = v.len();
        for _ in 0..v.len() {
            let top = v[0];
            v[0] = v[size - 1];
            v[size - 1] = top;
            size -= 1;
            Heap::move_down(v, size);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Heap;

    #[test]
    fn heap_operations() {
        let mut heap = Heap::new();
        assert!(heap.is_empty());

        heap.push(2);
        assert_eq!(heap.top(), 2);

        heap.push(5);
        assert_eq!(heap.top(), 5);

        heap.push(3);
        assert_eq!(heap.top(), 5);

        heap.pop();
        assert_eq!(heap.top(), 3);
    }

    #[test]
    fn heap_sort() {
        let mut v = vec![9, 3, 8, 10, 2, 4];
        Heap::sort(&mut v);
        assert_eq!(v, vec![2, 3, 4, 8, 9, 10])
    }
}
